FEAT: Log Setup

tomerpeceniski opened the pull request:
# FEAT: Log Setup

## Description
This PR creates a Log system for both User and Gym routes.
It logs information about the request, ongoing actions in the service layer and errors. It includes also request ID, for tracking all logs related with the same request.

## Checklist
- [X] All acceptance criteria have been met
- [X] My code follows the project's coding style.
- [ ] I have documented my code
- [ ] I wrote unit tests for the changes that I made

## Task in Notion
https://www.notion.so/Include-logs-on-the-system-172b8830dc7b80219344e3349ec1a304?pvs=4

tomerpeceniski committed:
update to main

tomerpeceniski committed:
feat: merge new features from main branch

tomerpeceniski committed:
feat: create different configuration profiles

tomerpeceniski committed:
feat: set up logger and create logging for errors

tomerpeceniski committed:
feat: set up logging in the user service layer

tomerpeceniski committed:
feat: create request logging interceptor for logging requests and reponses

tomerpeceniski committed:
refactor: hide path variables in request logs

tomerpeceniski committed:
chore(config): remove dev and prod profile files and consolidate into a single configuration

tomerpeceniski committed:
feat: add request id in logger stack

tomerpeceniski committed:
chore: remove period from API error message for consistency

tomerpeceniski committed:
feat: create logger for gym route

tomerpeceniski committed:
Merge branch 'main' into feat/log-setup

tomerpeceniski committed:
chore: change findAll method annotation due to non used parameter

danssaul commented on src/main/resources/application.properties:
```diff
@@ -6,7 +6,9 @@ spring.datasource.password=1234567
 spring.jpa.hibernate.ddl-auto=update
-spring.jpa.show-sql=true
-spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.show-sql=false
+spring.jpa.properties.hibernate.format_sql=false
```
this config is very useful to us, nice adition

danssaul commented on src/main/java/igym/services/GymService.java:
```diff
@@ -15,6 +17,7 @@
 @Service
 public class GymService {
```
considerer to use different log levels in which layer. In the service layer "info" for public methods, "debug" for decisions.

danssaul commented on src/main/java/igym/services/GymService.java:
```diff
@@ -23,27 +26,33 @@ public GymService(GymRepository gymRepository) {
+        logger.info("New gym created with id {}", savedGym.getId());
+        return savedGym;
     }
 
```
`

              @Transactional
                  public Gym createGym(Gym gym) {
                      logger.info("Attempting to create new gym with name {}", gym.getName());
                      logger.debug("Checking existence of gym name '{}'", gym.getName());
                      if (gymRepository.existsByName(gym.getName())) {
                          logger.warn("Gym creation failed - name '{}' already exists", gym.getName());
                          throw new DuplicateGymException("A gym with the name '" + gym.getName() + "' already exists");
                      }
                      logger.trace("Saving gym: {}", gym);
                      Gym savedGym = gymRepository.save(gym);
                      logger.info("New gym created with id {}", savedGym.getId());
                      return savedGym;
                  }
`

danssaul replied:
To works well with the new logs, I think we should update our logback.xml with new configurations per layer maybe:     <logger name="igym.services" level="DEBUG" />

danssaul commented on src/main/resources/application.properties:
```diff
@@ -6,7 +6,9 @@ spring.datasource.password=1234567
+spring.jpa.show-sql=false
+spring.jpa.properties.hibernate.format_sql=false
+
+logging.config=classpath:logback.xml
```
To use simple logs config I think the properties file is enough but since you are using an additional log file configuration considered to add colors to a log levels following the spring documentation [Color-coded Output](https://docs.spring.io/spring-boot/reference/features/logging.html#features.logging.log4j2-extensions.profile-specific:~:text=If%20your%20terminal%20supports)

danssaul commented on src/main/java/igym/controllers/UserController.java:
```diff
@@ -33,7 +33,7 @@ public UserController(UserService service) {
     @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<List<User>> findAll() {
+    public ResponseEntity<List<User>> findAll() {   
         List<User> list = service.findAll();
         return ResponseEntity.ok().body(list);
```
```suggestion
        List<User> list = service.findAll();
        logger.debug("Returned gym list size: {}", gyms.size());
        return ResponseEntity.ok().body(list);
```

danssaul reviewed (approved):
Good Job, I would like to see more logs levels and considerer in this PR including log profiles it will be helpful on our deploy. See [profile configuration](https://docs.spring.io/spring-boot/reference/features/logging.html#features.logging.log-groups:~:text=Copied!-,Profile%2Dspecific%20Configuration,-The%20%3CspringProfile%3E) in spring documentation

tomerpeceniski committed:
refactor: improve log visibility with debug level and custom formatting

- Added toString() methods in entities to enable detailed object logging
- Updated logback.xml to set proper log levels (DEBUG for app, INFO for framework)
- Added debug-level logs across application

tomerpeceniski committed:
feat: add dev and prod profiles with separate log levels

- Added application-dev.properties and application-prod.properties
- Created logback-dev.xml and logback-prod.xml for custom logging
- Activated dev profile by default for local development

tomerpeceniski committed:
feat: add color property for log levels

tomerpeceniski commented:
@daniel1sender hey, I updated the code after your review. Can you take a look on it?

danssaul commented on src/main/resources/application.properties:
```diff
@@ -1,12 +1,17 @@
+spring.jpa.show-sql=false
+spring.jpa.properties.hibernate.format_sql=false
+
+logging.config=classpath:logback.xml
```
I don't understood very well this line. For each file this line is referring?

danssaul reviewed (approved):
Nice, your profile configuration will be useful for our deploy!

tomerpeceniski committed:
fix: delete inexistent file reference

tomerpeceniski committed:
Merge branch 'main' into feat/log-setup

tomerpeceniski committed:
fix: delete duplicate method

tomerpeceniski committed:
fix: fix error messagein test  to correspond new pattern

danssaul commented on src/main/java/igym/entities/User.java:
```diff
@@ -49,4 +49,13 @@ public Status getStatus() {
         this.status = status;
     }
+
+    @Override
+    public String toString() {
+        return "{" +
+                "id=" + (id != null ? id : "null") +
+                ", name='" + name + '\'' +
+                ", status=" + status +
+                '}';
+    }
```
I don't understand why this method is defined here. where you are using it? and why? The Sonar is saying that is the cause of the an error. If you really need this method considerer to use lombok:
`
import lombok.ToString;

@ToString
public class Gym {
    ...
}
`

tomerpeceniski replied:
We need to have a toString method to log the entity. So we are using it when we log all entity data in our logs. I will go with your suggestion, it worked!

tomerpeceniski committed:
refactor: adjust code to meet CI and SonarQube quality rules

tomerpeceniski committed:
refactor: apply Lombok to entities and fix logger method

tomerpeceniski committed:
Feat: Add logs in update Gym endpoint
